Add to public url toggle create short link section

Coteh opened the pull request:
**Edit:** Toggle component has now been split into #534. This PR will now primarily be for the public URL toggle itself, which will use the Toggle component introduced in #534.

## New Behavior
### Description
- `isPublic` parameter added to `createURL` mutation request, when toggle is enabled it will be `true`, `false` otherwise (confirmed working on Firefox)
- ~Added `Subsection` component to house the new "Options" section.~ edit: no longer needed, see new comments

Components originally part of this PR then added to #534 instead:
- Added `Toggle` component, with an `onClick` handler.
- Toggle animations (confirmed working on Chrome, Firefox, and Safari)

### Known Issues
- ~The new subsection will stretch out the mobile view.~ edit: no longer the case, and `Subsection` has been removed from this PR

### Screenshots
![Screen Shot 2020-04-03 at 10 34 49 PM](https://user-images.githubusercontent.com/3276350/78417267-67b0bb80-75fe-11ea-9804-029a52d18339.png)
![Screen Shot 2020-04-03 at 10 45 17 PM](https://user-images.githubusercontent.com/3276350/78417268-6a131580-75fe-11ea-8095-137a09cfe1e6.png)
#### Old Screenshots
![toggle](https://user-images.githubusercontent.com/3276350/75086855-e21df400-5506-11ea-87aa-7b1742e6be6e.gif)
(with the old Toggle, see #534 for the new one)

byliuyang commented:
Let's divide this PR into the following ones: 
1) Add toggle button component
2) Add a new row in `New Short Link` `Section`.

byliuyang commented:
I wonder whether we can update the style of the button to like this: 
![image](https://user-images.githubusercontent.com/3537801/75850250-b1a64780-5d9b-11ea-995b-74575f60749e.png)

byliuyang commented:
Overall good job!!!

Coteh commented:
Updated style of toggle slider. Preview below:

![toggle_preview](https://user-images.githubusercontent.com/3276350/76136663-e1568900-6001-11ea-9399-378dd4c46e98.gif)

Slow-mo:

![toggle_slow](https://user-images.githubusercontent.com/3276350/76136673-fe8b5780-6001-11ea-9c72-84fa65949f06.gif)

byliuyang commented:
Fantastic work! Let's split out ToggleButton into a separate  PR so that we can get it checked in as soon as possible!

Coteh commented:
In addition to pulling in all recent changes from `master` (including the separated `Toggle` component which was added in #534 instead of this PR), I also added these new changes:
- Removed Subsection component
- Added a `margin-right` property to `Toggle.scss` which mirrors the `margin-left`, makes the toggle appear normal on mobile when not toggled
- Improved the spacing of the public URL toggle text
- Improved the spacing of the public URL toggle label for mobile

Coteh commented:
Some screenshots:

![Screen Shot 2020-04-03 at 10 34 49 PM](https://user-images.githubusercontent.com/3276350/78417267-67b0bb80-75fe-11ea-9804-029a52d18339.png)
![Screen Shot 2020-04-03 at 10 45 17 PM](https://user-images.githubusercontent.com/3276350/78417268-6a131580-75fe-11ea-8095-137a09cfe1e6.png)

byliuyang commented on frontend/src/component/pages/Home.tsx:
```diff
@@ -290,6 +297,7 @@ export class Home extends Component<Props, State> {
             onLongLinkTextFieldChange={this.handlerLongLinkChange}
             onShortLinkTextFieldBlur={this.handlerCustomAliasTextFieldBlur}
             onShortLinkTextFieldChange={this.handleAliasChange}
+            onPublicToggle={this.handlePublicToggle}
```
```suggestion
            onPublicToggle={this.handlePublicToggle}
```
```suggestion
            onPublicToggleClick={this.handlePublicToggleClick}
```

byliuyang commented on frontend/src/component/pages/Home.tsx:
```diff
@@ -290,6 +297,7 @@ export class Home extends Component<Props, State> {
             onLongLinkTextFieldChange={this.handlerLongLinkChange}
             onShortLinkTextFieldBlur={this.handlerCustomAliasTextFieldBlur}
```
This is not related to you. We need to change `handler` to `handle` in the future for code consistency.

Coteh replied:
TODO
- [ ] Change names of "handler" functions to say "handle" instead to be more consistent

byliuyang commented on frontend/src/component/pages/shared/CreateShortLinkSection.tsx:
```diff
@@ -18,6 +19,7 @@ interface Props {
   onLongLinkTextFieldChange?: (newLongLink: string) => void;
   onShortLinkTextFieldBlur?: () => void;
   onShortLinkTextFieldChange?: (newAlias: string) => void;
+  onPublicToggle?: (enabled: boolean) => void;
```
```suggestion
  onPublicToggleClick?: (enabled: boolean) => void;
```

byliuyang commented on frontend/src/component/pages/Home.tsx:
```diff
@@ -63,6 +63,7 @@ interface State {
   qrCodeUrl?: string;
   err?: IErr;
   inputErr?: string;
+  isPublic?: boolean;
```
```suggestion
  isShortLinkPublic?: boolean;
```

byliuyang commented on frontend/src/component/pages/Home.tsx:
```diff
@@ -244,6 +245,12 @@ export class Home extends Component<Props, State> {
       });
   };
 
+  handlePublicToggle = (enabled: boolean) => {
```
```suggestion
  handlePublicToggleClick = (enabled: boolean) => {
```

byliuyang commented on frontend/src/component/pages/shared/CreateShortLinkSection.tsx:
```diff
@@ -59,6 +61,12 @@ export class CreateShortLinkSection extends Component<Props> {
         </div>
         <div className={'input-error'}>{this.props.inputErr}</div>
+        <div>
+          <Toggle onClick={this.props.onPublicToggle}>
+            Make this short link public?
+            <br />(&nbsp;will be publicly listed&nbsp;)
+          </Toggle>
```
Few questions here:
- I wonder whether we can separate toggle description from the `Toggle` button itself so that we can leave developers flexibility for customization
- Instead of having extract description inside the parentheses,  can we make the description explicit to users? For example, change to 
```html
<!-- 
Margin right on desktop, full width on mobile. 
public-listing-toggle has border or background color.
-->
<div class="public-listing-toggle">Shared on <a>public feed</a>
    <Toggle onClick={this.props.onPublicToggle}></Toggle>
</div>
``` 
<img width="1069" alt="Screen Shot 2020-04-03 at 8 24 09 PM" src="https://user-images.githubusercontent.com/3537801/78417664-3da0ce80-75e9-11ea-80f6-4fb072873396.png">

The screenshot is just an example.

Coteh replied:
I like both those ideas, will make the toggle more clear and concise and the component itself will be more flexible. 👍

Coteh replied:
TODO
- [ ] Change label to "Shared on <a>public feed</a>"
- [ ] Separate label subcomponent from `Toggle` component, have the `Toggle` be a child of it
- [ ] Change style of the label subcomponent, play with border and background color

byliuyang commented on frontend/src/component/ui/Toggle.scss:
```diff
@@ -12,7 +13,11 @@ $toggle-shrink-size: 4px;
   justify-content: space-between;
   align-items: center;
-  width: 50%;
+  width: 60%;
```
Just wondering. What is this width for?

Coteh replied:
It's the size of the React component's top-level `div` element. (Which encompasses both the label and the button currently)

![image](https://user-images.githubusercontent.com/3276350/79054574-1911ab80-7c14-11ea-859a-af78ca13ec56.png)

Coteh replied:
If you increase the width to say `80%`, the toggle will be further away from the label.

Coteh replied:
With my latest commit, the `width` property is no longer a part of the `Toggle` component. This should hopefully allow the toggle to be easily added to more sections of the site.

byliuyang replied:
@Coteh Amazing! Thank you!!!

byliuyang replied:
@Coteh I like the way you think!

byliuyang commented on frontend/src/service/Url.service.ts:
```diff
@@ -134,12 +134,20 @@ export class UrlService {
     const captchaResponse = await this.captchaService.execute(
       CREATE_SHORT_LINK
     );
     let alias = link.alias === '' ? null : link.alias!;
```
I know this is not your code but let's help simplify:
```suggestion
```

byliuyang commented on frontend/src/service/Url.service.ts:
```diff
@@ -134,12 +134,20 @@ export class UrlService {
     );
     let alias = link.alias === '' ? null : link.alias!;
-    let variables = this.gqlCreateURLVariable(captchaResponse, link, alias);
+    let variables = this.gqlCreateURLVariable(
+      captchaResponse,
+      link,
+      alias,
+      isPublic
+    );
```
```suggestion
    let variables = this.gqlCreateURLVariable(
      captchaResponse,
      link,
      isPublic
    );
```

byliuyang commented on frontend/src/service/Url.service.ts:
```diff
@@ -134,12 +134,20 @@ export class UrlService {
   }
 
-  private async invokeCreateShortLinkApi(link: Url): Promise<Url> {
+  private async invokeCreateShortLinkApi(
+    link: Url,
+    isPublic: boolean
+  ): Promise<Url> {
```
```suggestion
  private async invokeCreateShortLinkApi(
    shortLink: Url,
    isPublic: boolean
  ): Promise<Url> {
```

Coteh commented:
Addressed some of the PR comments, and also fixed a style leak that resulted from a fix to another style leakage (https://github.com/short-d/short/pull/626/commits/69b99d5efc3bba74ca74bc055c8903cc065772a0)

Will check out the rest of the PR comments in a bit.

Coteh commented:
@byliuyang Separated the label (toggle description) from the `Toggle` component and also removed the `width` property from `Toggle`'s style. This should hopefully make the component more modular. Also switched positions of the label and toggle around, so toggle is on the left and label is on the right, but I can easily change it back. Here's how it looks now: (also added a border!)

![Screen Shot 2020-04-11 at 6 02 40 PM](https://user-images.githubusercontent.com/3276350/79056058-83304d80-7c20-11ea-9005-0c8de15fdf1c.png)

![Screen Shot 2020-04-11 at 6 02 54 PM](https://user-images.githubusercontent.com/3276350/79056060-862b3e00-7c20-11ea-8ee6-7ede22a40fb6.png)

byliuyang commented on frontend/src/service/Url.service.ts:
```diff
@@ -190,15 +196,14 @@ export class UrlService {
       urlInput: {
         originalURL: link.originalUrl,
-        customAlias: alias
+        customAlias: link.alias === '' ? null : link.alias!
```
Let's move this logic up and assign it a dedicated variable. Ideally avoid `ternary operator` when possible.

Coteh replied:
TODO
- [ ] Assign `link.alias === '' ? null : link.alias!` to its own variable to avoid use of ternaries

byliuyang commented on frontend/src/component/pages/shared/CreateShortLinkSection.tsx:
```diff
@@ -59,6 +61,12 @@ export class CreateShortLinkSection extends Component<Props> {
           </div>
         </div>
         <div className={'input-error'}>{this.props.inputErr}</div>
+        <div className={'public-listing-toggle'}>
```
I rethought about this. If we change the full border to a top divider ( border-top ), it may look cooler! Because the backend is not ready yet, we should use feature flag to turn it off ( hide it from homepage ) and turn it back on when it's ready.

You can move `<div className={'public-listing-toggle'}>` to [here](https://github.com/short-d/short/blob/master/frontend/src/component/UIFactory.tsx) and create a static feature flag [here](https://github.com/short-d/short/blob/master/frontend/src/service/StaticConfigDecision.service.ts)

Coteh replied:
TODO
- [ ] Move `public-listing-toggle` element to `UIFactory`
- [ ] Change border to just border-top
- [ ] Add feature flag to hide the public listing toggle from homepage

Coteh commented:
TODO (from pair session with @byliuyang ):
- [ ] Move divider out from the `PublicLinkToggle` component into `CreateShortLinkSection`
- [ ] Update `PublicLinkToggle` component to pass in the following props:
  - [ ] label
  - [ ] default state

Coteh commented:
Updated public listing toggle to reflect for fixes introduced in #693. Also generalized this component so that it can be used for not just future toggles in the front page, but also as toggles that may possibly be needed in other sections of short in the future such as admin panel.

@byliuyang and I discussed the option of instead combining the public listing toggle into the `CreateShortLinkSection` directly, which was also a valid solution due to the fact that no other toggleable business logic currently exists. However, I chose splitting it up into a separate component early as we need to enable/disable it with feature flag and it's much easier to do this when it's separated.

Screenshot of recent changes:
![Screen Shot 2020-04-26 at 1 28 30 PM](https://user-images.githubusercontent.com/3276350/80317734-83f5e180-87d3-11ea-8483-9bab009dfbe4.png)

Coteh commented:
Took a quick look at the code climate, seems like we can possibly refactor the way we call `UIFactory` functions, but I think it's beyond the scope of PR. What do you think we should do about this @byliuyang?

Ready for review.

byliuyang commented:
@Coteh I marked them as won't fix for now. Will look into it in the future.

Coteh commented:
Additional TODOs for myself:
- [ ] Use `Partial` construct as seen in `UserShortLinkSection` to assign default property for `PreferenceToggle`.

byliuyang commented on frontend/src/component/UIFactory.tsx:
```diff
@@ -130,6 +138,17 @@ export class UIFactory {
     );
   }
 
+  public createPublicListingToggle(props: any): ReactElement {
+    let ToggleLabel = (
+      <span>
+        Share on <span>public feed</span>
+      </span>
+    );
+    return (
+      <this.ToggledPublicListingToggle toggleLabel={ToggleLabel} {...props} />
+    );
+  }
```
This should be a separate component.

Coteh replied:
The toggle label you mean right?

byliuyang replied:
@Coteh The toggle and the label together.

Coteh replied:
I believe I already did this, it's called `PreferenceToggle`. It encapsulates a `Toggle` and a label.

byliuyang replied:
@Coteh Sorry, what I mean is to create `PublicListingToggle`

Coteh replied:
Right now a `PublicListingToggle` is just a toggle with a label that triggers an `onClick` function we pass down (from the `Home` component). I don't see any behaviour for `PublicListingToggle` that makes it any different from any other `PreferenceToggle`. Therefore, I think we should either keep it as `PreferenceToggle`, or rename `PreferenceToggle` to `PublicListingToggle` since we don't currently use `PreferenceToggle` for anything else at the moment.

byliuyang replied:
@Coteh Let's rename `PreferenceToggle` to `PublicListingToggle` and hard code the label inside.

Coteh replied:
Done

byliuyang commented on frontend/src/component/pages/shared/PreferenceToggle.tsx:
```diff
@@ -0,0 +1,32 @@
+export class PreferenceToggle extends Component<IProps> {
+  static defaultProps: Partial<IProps> = DEFAULT_PROPS;
+
+  private renderLabel() {
+    return <>{this.props.toggleLabel}</>;
+  }
```
```suggestion
```

Coteh replied:
Done

byliuyang commented on frontend/src/component/pages/shared/PreferenceToggle.tsx:
```diff
@@ -0,0 +1,32 @@
+
+  render() {
+    return (
+      <div className={'creation-toggle'}>
```
```suggestion
      <div className={'preference-toggle'}>
```

Coteh replied:
Done

byliuyang commented on frontend/src/component/pages/shared/PreferenceToggle.tsx:
```diff
@@ -0,0 +1,32 @@
+          defaultIsEnabled={this.props.defaultIsEnabled}
+          onClick={this.props.onToggleClick}
+        ></Toggle>
+        <span className={'toggle-label'}>{this.renderLabel()}</span>
```
<div>{this.props.toggleLabel}</div>

Please use flex layout for `preference-toggle`.

Coteh replied:
I think it already is... 
![image](https://user-images.githubusercontent.com/3276350/80824948-6b0f7680-8bad-11ea-8398-005cf6240652.png)

byliuyang commented on frontend/src/service/Url.service.ts:
```diff
@@ -179,15 +185,15 @@ export class UrlService {
-    alias: string | null,
     isPublic: boolean = false
   ) {
+    let linkAlias = link.alias === '' ? null : link.alias!;
```
```suggestion
    let alias = link.alias === '' ? null : link.alias!;
```

Coteh replied:
Done

byliuyang commented on frontend/src/service/Url.service.ts:
```diff
@@ -179,15 +185,15 @@ export class UrlService {
       urlInput: {
         originalURL: link.originalUrl,
-        customAlias: alias
+        customAlias: linkAlias
```
```suggestion
        customAlias: alias
```

byliuyang replied:
I wonder why can't we directly pass alias?

Coteh replied:
This is what happens when you specify an empty custom alias and pass it directly into `urlInput` argument:

![image](https://user-images.githubusercontent.com/3276350/80825439-66978d80-8bae-11ea-8413-2cc387898240.png)

Looks something like this in `url` table in DB:

![image](https://user-images.githubusercontent.com/3276350/80825473-78793080-8bae-11ea-8714-f68980e53d48.png)

If we instead pass `null` as the argument, this behaviour will be prevented and an unknown error will be thrown instead:

![image](https://user-images.githubusercontent.com/3276350/80825580-afe7dd00-8bae-11ea-8fac-d926d9c54bf4.png)

Should we instead throw an explicit error when user passes in an empty custom alias? That would probably be the best way to deal with this case I think.

byliuyang replied:
@Coteh I think we need to fix the backend alias validation.

byliuyang replied:
Let's open another PR to fix the backend!

byliuyang replied:
I see. This code exists before your changes.

Let's add a TODO to fix the backend validation code. We can't always trust the input from the frontend because malicious users can use JavaScript to modify our page: https://martinfowler.com/articles/web-security-basics.html. Please read `Trust` and `Input Validation` part. At the backend, we should treat empty alias as automatic alias.

Coteh replied:
Issue created https://github.com/short-d/short/issues/728

byliuyang commented on frontend/src/component/pages/shared/PreferenceToggle.spec.tsx:
```diff
@@ -0,0 +1,57 @@
+import { IPagedShortLinks } from '../../../service/ShortLink.service';
+
+describe('PreferenceToggle component', () => {
+  const TOGGLE_CLASS_NAME = '.toggle';
+  const TOGGLE_SELECTOR = TOGGLE_CLASS_NAME;
+  const TOGGLE_ACTIVE_SELECTOR = TOGGLE_CLASS_NAME + '> .active';
```
let's move the setup code into the test themselves so that each test is independent from each other. It's okey to have repeatitation here.

byliuyang replied:
https://github.com/short-d/short#the-first-principal

Coteh replied:
Done

byliuyang commented on frontend/src/service/Url.service.ts:
```diff
@@ -179,9 +185,9 @@ export class UrlService {
-    alias: string | null,
     isPublic: boolean = false
   ) {
+    let alias = link.alias === '' ? null : link.alias!;
```
This part of business logic. I think it should happen outside this function. I wonder if we can remove this completely.

byliuyang commented on frontend/src/component/pages/shared/PreferenceToggle.tsx:
```diff
@@ -0,0 +1,28 @@
+          defaultIsEnabled={this.props.defaultIsEnabled}
+          onClick={this.props.onToggleClick}
+        ></Toggle>
+        <span className={'toggle-label'}>{this.props.toggleLabel}</span>
```
```suggestion
        <div className={'toggle-label'}>{this.props.toggleLabel}</div>
```

byliuyang replied:
Changing to div requires us to use flex layout on `preference-toggle`

Coteh replied:
Already made `preference-toggle` a `flex`, but I have now moved it into its own `scss` file which I will push momentarily.

byliuyang commented:
We are almost there!

byliuyang commented on frontend/src/service/Url.service.ts:
```diff
@@ -179,9 +185,9 @@ export class UrlService {
-    alias: string | null,
     isPublic: boolean = false
   ) {
+    let alias = link.alias === '' ? null : link.alias!;
```
Please add a TODO to mention the backend change we have to make.

byliuyang replied:
Just saw the issue being opened. This one is resolved.

Coteh replied:
Done

byliuyang commented on frontend/src/component/pages/shared/PublicListingToggle.tsx:
```diff
@@ -0,0 +1,30 @@
+          defaultIsEnabled={this.props.defaultIsEnabled}
+          onClick={this.props.onToggleClick}
+        ></Toggle>
+        <span className={'toggle-label'}>
```
Let's change this `span` to `div`.

Coteh replied:
Done

byliuyang commented on frontend/src/component/pages/shared/PublicListingToggle.tsx:
```diff
@@ -0,0 +1,30 @@
+          onClick={this.props.onToggleClick}
+        ></Toggle>
+        <span className={'toggle-label'}>
+          Share on <span>public feed</span>
```
```suggestion
          Share on <a herf="/public" target="_blank">public feed</a>
```

Coteh replied:
Done

byliuyang commented on frontend/src/component/pages/shared/CreateShortLinkSection.scss:
```diff
@@ -12,3 +13,7 @@
   margin: 10px;
 }
+
+.preference-toggles > .preference-toggle:first-child {
```
```suggestion
.preference-toggles {
```

byliuyang replied:
I think we are separating the two subsections.

Coteh replied:
Yeah it's definitely not needed anymore. Done
